src/wasm2asm.h

```diff
@@ -1696,38 +1835,55 @@ Ref Wasm2AsmBuilder::processFunctionBody(Function* func, IString result) {
       return ValueBuilder::makeCall(ABORT_FUNC);
     }
   };
-  return ExpressionProcessor(this, func).visit(func->body, result);
+  return ExpressionProcessor(this, m, func).visit(func->body, result);
 }
 
-static void makeInstantiation(Ref ret) {
-  // var __array_buffer = new ArrayBuffer(..)
-  Ref mem = ValueBuilder::makeNew(
-      ValueBuilder::makeCall(ARRAY_BUFFER, ValueBuilder::makeInt(0x10000)));
-  Ref arrayBuffer = ValueBuilder::makeVar();
+static void makeInstantiation(Ref ret, Name funcName, Name moduleName, bool first) {
   Name buffer("__array_buffer");
-  ValueBuilder::appendToVar(arrayBuffer, buffer, mem);
-  flattenAppend(ret, arrayBuffer);
-
-  // var HEAP32 = new Int32Array(__array_buffer);
-  Ref heap32Array = ValueBuilder::makeNew(
-      ValueBuilder::makeCall(INT32ARRAY, ValueBuilder::makeName(buffer)));
-  Ref heap32 = ValueBuilder::makeVar();
-  ValueBuilder::appendToVar(heap32, HEAP32, heap32Array);
-  flattenAppend(ret, heap32);
-
-  // var HEAPF32 = new Float32Array(__array_buffer);
-  Ref heapf32Array = ValueBuilder::makeNew(
-      ValueBuilder::makeCall(FLOAT32ARRAY, ValueBuilder::makeName(buffer)));
-  Ref heapf32 = ValueBuilder::makeVar();
-  ValueBuilder::appendToVar(heapf32, HEAPF32, heapf32Array);
-  flattenAppend(ret, heapf32);
-
-  // var HEAPF64 = new Float64Array(__array_buffer);
-  Ref heapf64Array = ValueBuilder::makeNew(
-      ValueBuilder::makeCall(FLOAT64ARRAY, ValueBuilder::makeName(buffer)));
-  Ref heapf64 = ValueBuilder::makeVar();
-  ValueBuilder::appendToVar(heapf64, HEAPF64, heapf64Array);
-  flattenAppend(ret, heapf64);
+  if (first) {
+    flattenAppend(ret, ValueBuilder::makeName(R"(
+      var __array_buffer = new ArrayBuffer(65536)
+      var HEAP32 = new Int32Array(__array_buffer);
+      var HEAPF32 = new Float32Array(__array_buffer);
+      var HEAPF64 = new Float64Array(__array_buffer);
+      var nan = NaN;
+      var infinity = Infinity;
+    )"));
+
+    // When equating floating point values in spec tests we want to use bitwise
+    // equality like wasm does. Unfortunately though NaN makes this tricky. JS
+    // implementations like Spidermonkey and JSC will canonicalize NaN loads from
+    // `Float32Array`, but V8 will not. This means that NaN representations are
+    // kind of all over the place and difficult to bitwise equate.
+    //
+    // To work around this problem we just use a small shim which considers all
+    // NaN representations equivalent and otherwise tests for bitwise equality.
+    flattenAppend(ret, ValueBuilder::makeName(R"(
```

Comment:
Indeed! And I'd be totally fine using some more official!

Reply:
I see. And meanwhile I see this already existed before, it was just moved around.

My concern is what this might do to optimization possibilities. Does this get emitted alongside the compiled functions? In general, I was thinking that we'd want to run an asm.js optimizer on those, even if the code as a whole is not asm.js. Stuff that is not valid asm.js could be imported from the outside or something like that. Anyhow, if it's just a few specific functions we can probably just special-case them.

But this does make me think that perhaps we should not remove asm.js validation as quickly as we were thinking. If it's practical to keep the "inner" portion valid asm.js, that would make optimizing it easier.

Reply:
FWIW this code injected here is purely for the test harness, it is only used in `--allow-asserts` mode which is basically internal testing as I understood it?

Reply:
I think so, yeah.

Reply:
If that's the case though the optimization possibilities aren't too improtant, right?

Reply:
Not from this specifically, but in general if we move to non-asm.js then the question arises there. I commented to the general planning issue with some thoughts.

Reply:
Ok, do you want me to change this to something else in the meantime?

Reply:
No, thanks, I think it's ok for now.